src/bin/inx-chronicle/stardust_inx/chunks.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Mutex,
+};
+
+pub trait ChunksExt: Iterator {
+    fn chunks(self, size: usize) -> IntoChunks<Self>
+    where
+        Self: Sized,
+    {
+        IntoChunks {
+            inner: Mutex::new(GroupInner {
+                key: ChunkIndex::new(size),
+                iter: self,
+                current_key: None,
+                current_elt: None,
+                done: false,
+                top_group: 0,
+                oldest_buffered_group: 0,
+                bottom_group: 0,
+                buffer: Vec::new(),
+                dropped_group: !0,
+            }),
+            index: AtomicUsize::new(0),
+        }
+    }
+}
+impl<T: Iterator> ChunksExt for T {}
+
+pub struct IntoChunks<I: Iterator> {
+    inner: Mutex<GroupInner<usize, I, ChunkIndex>>,
+    index: AtomicUsize,
+}
+
+impl<I: Iterator> IntoChunks<I> {
+    /// `client`: Index of chunk that requests next element
+    fn step(&self, client: usize) -> Option<I::Item> {
+        self.inner.lock().unwrap().step(client)
+    }
+
+    /// `client`: Index of chunk
+    fn drop_group(&self, client: usize) {
+        self.inner.lock().unwrap().drop_group(client)
+    }
+}
+
+impl<'a, I: Iterator> IntoIterator for &'a IntoChunks<I> {
+    type Item = Chunk<'a, I>;
+    type IntoIter = Chunks<'a, I>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Chunks { parent: self }
+    }
+}
+
+pub struct Chunks<'a, I: Iterator> {
+    parent: &'a IntoChunks<I>,
+}
+
+impl<'a, I: Iterator> Iterator for Chunks<'a, I> {
+    type Item = Chunk<'a, I>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.parent.index.fetch_add(1, Ordering::Relaxed);
+        self.parent.step(index).map(|elt| Chunk {
+            parent: self.parent,
+            index,
+            first: Some(elt),
+        })
+    }
+}
+
+pub struct Chunk<'a, I: Iterator> {
+    parent: &'a IntoChunks<I>,
+    index: usize,
+    first: Option<I::Item>,
+}
+
+impl<'a, I: Iterator> Drop for Chunk<'a, I> {
+    fn drop(&mut self) {
+        self.parent.drop_group(self.index);
+    }
+}
+
+impl<'a, I: Iterator> Iterator for Chunk<'a, I> {
+    type Item = I::Item;
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if let elt @ Some(..) = self.first.take() {
```

Comment:
ugh, that's ugly, haha. Isn't this basically the same as `if let Some(elt) = self.first.take()`?

Reply:
No, it binds `elt` to the `Option` not the interior value.

Reply:
ah, alright, now I see it!